File: src/signaling-server.py
#!/usr/bin/env python
#
# Python signaling server example for libdatachannel
# Copyright (c) 2020 Paul-Louis Ageneau
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.

import sys
import ssl
import json
import time
import asyncio
import logging
import websockets
from websockets.legacy.server import WebSocketServerProtocol, serve as ws_serve
from aiohttp import web


logger = logging.getLogger('websockets')
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler(sys.stdout))


clients = {}       # client_id -> websocket
device_info = {}  # client_id -> {connected_at, path, type, last_seen}
INACTIVITY_TIMEOUT = 90  # seconds - disconnect if no activity
KEEPALIVE_INTERVAL = 30  # seconds - how often we ping each client
KEEPALIVE_PING_TIMEOUT = 10  # seconds - how long we wait for a pong


def load_turn_config():
    try:
        with open('cfg/turn.conf', 'r') as f:
            servers = json.load(f)
            if isinstance(servers, list):
                return servers
    except Exception as e:
        print('Failed to read turn.conf: {}'.format(e))
    return []


def load_version():
    try:
        with open('cfg/version', 'r') as f:
            return f.read().strip()
    except Exception:
        return 'unknown'


async def keepalive_loop(websocket, client_id, interval=KEEPALIVE_INTERVAL, timeout=KEEPALIVE_PING_TIMEOUT):
    """Explicit ping/pong keepalive with logging.

    Replaces websockets' built-in automatic ping_interval/ping_timeout
    mechanism (which has no user-visible hook for logging received
    ping/pong frames) with an explicit per-client task so we can log RTT
    and detect a dead connection ourselves.
    """
    try:
        while True:
            await asyncio.sleep(interval)
            try:
                pong_waiter = await websocket.ping()
                start = time.time()
                await asyncio.wait_for(pong_waiter, timeout=timeout)
                rtt_ms = (time.time() - start) * 1000
                if client_id in device_info:
                    device_info[client_id]['last_seen'] = time.time()
                    device_info[client_id]['rtt'] = rtt_ms
            except asyncio.TimeoutError:
                print('[{}] pong timeout after {}s, closing'.format(client_id, timeout))
                await websocket.close(1000, 'Ping timeout')
                break
    except (asyncio.CancelledError, websockets.exceptions.ConnectionClosed):
        pass  # normal on disconnect/shutdown


async def handle_websocket(websocket):
    client_id = None
    keepalive_task = None
    try:
        print('New connection from {} path={}'.format(websocket.remote_address, websocket.path))
        splitted = websocket.path.split('/')
        splitted.pop(0)
        client_id = splitted.pop(0)
        websocket._client_id = client_id
        print('Client {} connected'.format(client_id))

        device_type = 'consumer' if client_id.startswith('Player') else 'producer'
        clients[client_id] = websocket
        device_info[client_id] = {
            'connected_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
            'path': websocket.path,
            'type': device_type,
            'last_seen': time.time(),
            'rtt' : 0
        }

        # Start explicit keepalive/RTT logging for this client
        keepalive_task = asyncio.create_task(keepalive_loop(websocket, client_id))

        while True:
            try:
                data = await websocket.recv()
            except websockets.exceptions.ConnectionClosed as e:
                print('[{}] Connection closed: code={}, reason={}'.format(client_id, e.code, e.reason))
                break
            if client_id in device_info:
                device_info[client_id]['last_seen'] = time.time()
            print('[{}] << {}'.format(client_id, data))
            message = json.loads(data)

            # Handle clean disconnect message
            if message.get('type') == 'bye':
                print('[{}] sent bye, closing cleanly'.format(client_id))
                await websocket.close(1000, 'Client disconnected')
                break

            destination_id = message['id']
            destination_websocket = clients.get(destination_id)
            if destination_websocket:
                message['id'] = client_id
                if client_id.startswith('Player') and message.get('type') == 'request':
                    servers = load_turn_config()
                    if servers:
                        message['turn'] = servers
                        print('[{}] TURN padded to request for {}'.format(client_id, destination_id))
                data = json.dumps(message)
                # print('[{}] >> {}'.format(destination_id, data))
                send_start = time.time()
                await destination_websocket.send(data)
                send_elapsed = (time.time() - send_start) * 1000
                print('[{}] >> {} : sent in {:.2f}ms'.format(destination_id, data, send_elapsed))
            else:
                print('Client {} not found'.format(destination_id))
                error_response = {
                    "type": "error",
                    "msg": "Device offline"
                }
                await websocket.send(json.dumps(error_response))

    except websockets.exceptions.WebSocketException as e:
        print('[{}] WebSocket error: {}'.format(client_id, e))
    except Exception as e:
        print('[{}] Unexpected error: {}'.format(client_id, e))

    finally:
        if keepalive_task:
            keepalive_task.cancel()
        if client_id:
            del clients[client_id]
            device_info.pop(client_id, None)
            print('Client {} disconnected'.format(client_id))


# ── HTTP REST handlers ──────────────────────────────────────────────────────

async def http_devices(request):
    """GET /devices — return all currently connected clients.
    Query params:
      type  — filter by device type (e.g. ?type=producer)
    """
    type_filter = request.query.get('type', None)
    result = []
    for cid, info in device_info.items():
        if type_filter and info.get('type') != type_filter:
            continue
        result.append({
            'id': cid,
            'connected_at': info['connected_at'],
            'path': info['path'],
            'type': info['type'],
            'rtt': info['rtt'],
        })
    return web.json_response({
        'count': len(result),
        'devices': result,
    }, headers={'Access-Control-Allow-Origin': '*'})


async def http_health(request):
    """GET /health — liveness check."""
    return web.json_response({
        'status': 'ok',
        'clients': len(clients),
        'version': load_version(),
    }, headers={'Access-Control-Allow-Origin': '*'})


# ── Background tasks ───────────────────────────────────────────────────────

async def sweep_inactive_clients():
    """Periodically disconnect clients that haven't sent pings or messages."""
    while True:
        await asyncio.sleep(30)
        now = time.time()
        stale = [cid for cid, info in device_info.items()
                 if now - info.get('last_seen', 0) > INACTIVITY_TIMEOUT]
        for cid in stale:
            ws = clients.get(cid)
            if ws:
                print('[{}] Inactive for {}s, disconnecting'.format(cid, INACTIVITY_TIMEOUT))
                await ws.close(1000, 'Inactivity timeout')


# ── Main ─────────────────────────────────────────────────────────────────────

async def main():
    # Usage: ./server.py [[host:]port] [SSL certificate file] [http port]
    # Example: ./server.py 8000 cert.pem 8080
    endpoint_or_port = sys.argv[1] if len(sys.argv) > 1 else "8000"
    http_port        = int(sys.argv[2]) if len(sys.argv) > 2 else 8080
    ssl_cert         = sys.argv[3] if len(sys.argv) > 3 else None

    endpoint = endpoint_or_port if ':' in endpoint_or_port else "127.0.0.1:" + endpoint_or_port

    if ssl_cert:
        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        ssl_context.load_cert_chain(ssl_cert)
    else:
        ssl_context = None

    host, port = endpoint.rsplit(':', 1)

    # WebSocket signaling server
    # Keepalive is handled explicitly per-client via keepalive_loop() above
    # (started in handle_websocket), so the built-in automatic ping is
    # disabled here to avoid double-pinging the same connection.
    ws_server = await ws_serve(
        handle_websocket,
        host,
        int(port),
        ssl=ssl_context,
        create_protocol=WebSocketServerProtocol,
        ping_interval=None,
        ping_timeout=None,
        close_timeout=10,
        max_queue=32,
        compression=None
    )

    # Start inactivity sweeper
    asyncio.create_task(sweep_inactive_clients())
    print('WebSocket signaling listening on {}'.format(endpoint))

    # HTTP REST server (always plain HTTP, sits behind nginx)
    app = web.Application()
    app.router.add_get('/devices', http_devices)
    app.router.add_get('/health',  http_health)
    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, host, http_port)
    await site.start()
    print('HTTP REST API listening on {}:{}'.format(host, http_port))
    print('  GET http://{}:{}/devices  — list online devices'.format(host, http_port))
    print('  GET http://{}:{}/health   — health check'.format(host, http_port))

    await ws_server.wait_closed()


if __name__ == '__main__':
    asyncio.run(main())

File: src/go/src/signaling-server.go
package main

import (
	"crypto/tls"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

const (
	InactivityTimeout    = 90 * time.Second
	KeepaliveInterval    = 30 * time.Second
	KeepalivePingTimeout = 10 * time.Second
	WriteWait            = 10 * time.Second
)

type DeviceInfo struct {
	ID          string    `json:"id"`
	ConnectedAt string    `json:"connected_at"`
	Path        string    `json:"path"`
	Type        string    `json:"type"`
	RTT         float64   `json:"rtt"`
	LastSeen    time.Time `json:"-"`
}

type Client struct {
	ID        string
	Conn      *websocket.Conn
	Info      *DeviceInfo
	Send      chan []byte
	done      chan struct{}
	closeOnce sync.Once
}

func (c *Client) Close() {
	c.closeOnce.Do(func() { close(c.done) })
}

type Server struct {
	clients    map[string]*Client
	mu         sync.RWMutex
	upgrader   websocket.Upgrader
	playerPath string
	rootDir    string
}

func NewServer(playerPath string, rootDir string) *Server {
	return &Server{
		clients: make(map[string]*Client),
		upgrader: websocket.Upgrader{
			ReadBufferSize:  1024,
			WriteBufferSize: 1024,
			CheckOrigin:     func(r *http.Request) bool { return true },
		},
		playerPath: playerPath,
		rootDir:    rootDir,
	}
}

func (s *Server) loadTurnConfig() []map[string]interface{} {
	data, err := os.ReadFile(s.rootDir + "cfg/turn.conf")
	if err != nil {
		return nil
	}
	var servers []map[string]interface{}
	if err := json.Unmarshal(data, &servers); err != nil {
		return nil
	}
	return servers
}

func (s *Server) loadVersion() string {
	data, err := os.ReadFile(s.rootDir + "cfg/version")
	if err != nil {
		return "unknown"
	}
	return strings.TrimSpace(string(data))
}

func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	switch {
	case r.URL.Path == "/":
		http.ServeFile(w, r, s.playerPath)
	case r.URL.Path == "/api/devices":
		s.handleDevices(w, r)
	case r.URL.Path == "/api/health":
		s.handleHealth(w, r)
	case strings.HasPrefix(r.URL.Path, "/downloads/"):
		s.handleDownloads(w, r)
	default:
		s.handleWebSocket(w, r)
	}
}

func (s *Server) handleWebSocket(w http.ResponseWriter, r *http.Request) {
	clientID := strings.Split(strings.TrimPrefix(r.URL.Path, "/"), "/")[0]
	if clientID == "" {
		http.Error(w, "Client ID required", http.StatusBadRequest)
		return
	}

	conn, err := s.upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Printf("Upgrade failed: %v", err)
		return
	}

	log.Printf("New connection from %s path=%s", r.RemoteAddr, r.URL.Path)

	deviceType := "consumer"
	if !strings.HasPrefix(clientID, "Player") {
		deviceType = "producer"
	}

	c := &Client{
		ID:   clientID,
		Conn: conn,
		Info: &DeviceInfo{
			ID:          clientID,
			ConnectedAt: time.Now().UTC().Format(time.RFC3339),
			Path:        r.URL.Path,
			Type:        deviceType,
			LastSeen:    time.Now(),
		},
		Send: make(chan []byte, 256),
		done: make(chan struct{}),
	}

	s.mu.Lock()
	if old, exists := s.clients[clientID]; exists {
		old.Close()
		old.Conn.Close()
	}
	s.clients[clientID] = c
	s.mu.Unlock()

	log.Printf("Client %s connected", clientID)

	go s.writePump(c)
	s.readPump(c)
}

func (s *Server) readPump(c *Client) {
	defer func() {
		s.mu.Lock()
		if s.clients[c.ID] == c {
			delete(s.clients, c.ID)
		}
		s.mu.Unlock()
		c.Conn.Close()
		c.Close()
		log.Printf("Client %s disconnected", c.ID)
	}()

	pongWait := KeepaliveInterval + KeepalivePingTimeout
	c.Conn.SetReadLimit(512 * 1024)
	c.Conn.SetReadDeadline(time.Now().Add(pongWait))
	c.Conn.SetPongHandler(func(string) error {
		c.Conn.SetReadDeadline(time.Now().Add(pongWait))
		c.Info.LastSeen = time.Now()
		return nil
	})

	for {
		_, message, err := c.Conn.ReadMessage()
		if err != nil {
			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseNormalClosure) {
				log.Printf("[%s] Read error: %v", c.ID, err)
			}
			break
		}

		c.Info.LastSeen = time.Now()
		log.Printf("[%s] << %s", c.ID, string(message))

		var msg map[string]interface{}
		if err := json.Unmarshal(message, &msg); err != nil {
			continue
		}

		if msg["type"] == "bye" {
			log.Printf("[%s] sent bye, closing cleanly", c.ID)
			c.Conn.WriteControl(websocket.CloseMessage,
				websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""),
				time.Now().Add(WriteWait))
			break
		}

		destID, _ := msg["id"].(string)
		if destID == "" {
			continue
		}

		s.mu.RLock()
		dest, exists := s.clients[destID]
		s.mu.RUnlock()

		if exists {
			msg["id"] = c.ID
			if strings.HasPrefix(c.ID, "Player") && msg["type"] == "request" {
				if servers := s.loadTurnConfig(); len(servers) > 0 {
					msg["turn"] = servers
				}
			}
			data, _ := json.Marshal(msg)
			select {
			case dest.Send <- data:
				log.Printf("[%s] >> %s", destID, string(data))
			default:
			}
		} else {
			data, _ := json.Marshal(map[string]string{"type": "error", "msg": "Device offline"})
			select {
			case c.Send <- data:
			default:
			}
		}
	}
}

func (s *Server) writePump(c *Client) {
	ticker := time.NewTicker(KeepaliveInterval)
	defer func() {
		ticker.Stop()
		c.Conn.Close()
	}()

	for {
		select {
		case msg, ok := <-c.Send:
			c.Conn.SetWriteDeadline(time.Now().Add(WriteWait))
			if !ok {
				c.Conn.WriteMessage(websocket.CloseMessage, []byte{})
				return
			}
			if err := c.Conn.WriteMessage(websocket.TextMessage, msg); err != nil {
				return
			}
		case <-ticker.C:
			c.Conn.SetWriteDeadline(time.Now().Add(WriteWait))
			if err := c.Conn.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}
		case <-c.done:
			return
		}
	}
}

func (s *Server) sweepInactiveClients() {
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()

	for range ticker.C {
		now := time.Now()
		s.mu.RLock()
		var stale []*Client
		for _, c := range s.clients {
			if now.Sub(c.Info.LastSeen) > InactivityTimeout {
				stale = append(stale, c)
			}
		}
		s.mu.RUnlock()

		for _, c := range stale {
			log.Printf("[%s] Inactive for %v, disconnecting", c.ID, InactivityTimeout)
			c.Close()
		}
	}
}

func (s *Server) handleDevices(w http.ResponseWriter, r *http.Request) {
	typeFilter := r.URL.Query().Get("type")
	s.mu.RLock()
	defer s.mu.RUnlock()

	devices := make([]DeviceInfo, 0)
	for _, c := range s.clients {
		if typeFilter != "" && c.Info.Type != typeFilter {
			continue
		}
		devices = append(devices, *c.Info)
	}

	w.Header().Set("Access-Control-Allow-Origin", "*")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"count":   len(devices),
		"devices": devices,
	})
}

func (s *Server) handleHealth(w http.ResponseWriter, r *http.Request) {
	s.mu.RLock()
	n := len(s.clients)
	s.mu.RUnlock()

	w.Header().Set("Access-Control-Allow-Origin", "*")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"status":  "ok",
		"clients": n,
		"version": s.loadVersion(),
	})
}

func (s *Server) handleDownloads(w http.ResponseWriter, r *http.Request) {
	filename := strings.TrimPrefix(r.URL.Path, "/downloads/")
	if filename == "" || strings.Contains(filename, "..") {
		http.Error(w, "Invalid filename", http.StatusBadRequest)
		return
	}
	filepath := s.rootDir + "downloads/" + filename
	http.ServeFile(w, r, filepath)
}

func main() {
	port := flag.Int("port", 8000, "Listen port")
	rootDir := flag.String("root", "", "Root directory for config and downloads (default: current dir)")
	https := flag.Bool("https", false, "Enable HTTPS (uses cert.pem from root directory)")
	flag.Parse()

	root := *rootDir
	if root != "" && !strings.HasSuffix(root, "/") {
		root += "/"
	}

	playerPath := root + "player/ws.html"
	srv := NewServer(playerPath, root)
	go srv.sweepInactiveClients()

	addr := fmt.Sprintf(":%d", *port)
	httpServer := &http.Server{
		Addr:    addr,
		Handler: srv,
	}

	if *https {
		certPath := root + "cert.pem"
		cert, err := tls.LoadX509KeyPair(certPath, certPath)
		if err != nil {
			log.Fatalf("TLS: %v", err)
		}
		httpServer.TLSConfig = &tls.Config{
			Certificates: []tls.Certificate{cert},
			MinVersion:   tls.VersionTLS12,
		}
		log.Printf("Listening on wss://0.0.0.0%s", addr)
		log.Printf("  GET https://0.0.0.0%s/         — WebRTC Player", addr)
		log.Printf("  GET https://0.0.0.0%s/api/devices  — list online devices", addr)
		log.Printf("  GET https://0.0.0.0%s/api/health   — health check", addr)
		log.Fatal(httpServer.ListenAndServeTLS("", ""))
	} else {
		log.Printf("Listening on ws://0.0.0.0%s", addr)
		log.Printf("  GET http://0.0.0.0%s/         — WebRTC Player", addr)
		log.Printf("  GET http://0.0.0.0%s/api/devices  — list online devices", addr)
		log.Printf("  GET http://0.0.0.0%s/api/health   — health check", addr)
		log.Fatal(httpServer.ListenAndServe())
	}
}
